File: backend/meta/indicator.py
"""
语义指标实体
Indicator是域级别的语义指标
"""
from typing import Optional
from datetime import date


class Indicator:
    """语义指标实体 - 域级别资源"""
    
    def __init__(
        self,
        id: int,
        name: str,
        expression: str,
        returnType: str,
        domainId: int,
        unit: Optional[str] = None,
        description: Optional[str] = None,
        status: str = "draft",
        createdAt: Optional[str] = None,
        updatedAt: Optional[str] = None
    ):
        self.id = id
        self.name = name
        self.expression = expression
        self.returnType = returnType
        self.domainId = domainId
        self.unit = unit
        self.description = description
        self.status = status  # draft, published, offline
        self.createdAt = createdAt or date.today().isoformat()
        self.updatedAt = updatedAt or date.today().isoformat()
    
    def to_dict(self) -> dict:
        """转换为字典"""
        return {
            "id": self.id,
            "name": self.name,
            "expression": self.expression,
            "returnType": self.returnType,
            "unit": self.unit,
            "description": self.description,
            "status": self.status,
            "domainId": self.domainId,
            "createdAt": self.createdAt,
            "updatedAt": self.updatedAt
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Indicator':
        """从字典创建实例"""
        return cls(
            id=data["id"],
            name=data["name"],
            expression=data.get("expression", ""),
            returnType=data.get("returnType", "number"),
            domainId=data["domainId"],
            unit=data.get("unit"),
            description=data.get("description"),
            status=data.get("status", "draft"),
            createdAt=data.get("createdAt"),
            updatedAt=data.get("updatedAt")
        )
    
    def publish(self) -> 'Indicator':
        """发布指标"""
        self.status = "published"
        self.updatedAt = date.today().isoformat()
        return self
    
    def offline(self) -> 'Indicator':
        """下线指标"""
        self.status = "offline"
        self.updatedAt = date.today().isoformat()
        return self
    
    def __repr__(self):
        return f"Indicator(id={self.id}, name='{self.name}', status='{self.status}')"

File: backend_java/src/main/java/com/jianmo/platform/entity/Indicator.java
package com.jianmo.platform.entity;

import com.jianmo.platform.common.BaseEntity;
import javax.persistence.*;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;

@Data
@EqualsAndHashCode(callSuper = true)
@NoArgsConstructor
@AllArgsConstructor
@Entity
@Table(name = "indicators")
public class Indicator extends BaseEntity {

    @Column(nullable = false, unique = true)
    private String code;

    @Column(nullable = false)
    private String name;

    @Column(name = "domain_id")
    private Long domainId;

    @Column(columnDefinition = "TEXT")
    private String description;

    /**
     * 计算表达式（如：SUM(amount) / COUNT(*)）
     */
    @Column(nullable = false, columnDefinition = "TEXT")
    private String expression;

    /**
     * 返回类型（number, string, boolean, date等）
     */
    @Column(name = "return_type")
    private String returnType = "number";

    /**
     * 单位（如：元、个、次等）
     */
    private String unit;

    /**
     * 维度字段（JSON格式存储）
     */
    @Column(columnDefinition = "TEXT")
    private String dimensions;

    /**
     * 过滤条件（JSON格式存储）
     */
    @Column(columnDefinition = "TEXT")
    private String filters;

    /**
     * 排序字段（JSON格式存储）
     */
    @Column(name = "sort_fields", columnDefinition = "TEXT")
    private String sortFields;

    /**
     * 关联的属性ID列表（JSON格式存储）
     */
    @Column(name = "related_properties", columnDefinition = "TEXT")
    private String relatedProperties;

    /**
     * 状态（draft, active, deprecated）
     */
    private String status = "draft";

    /**
     * 创建者
     */
    private String creator;

    public Indicator(String code, String name, String expression, Long domainId) {
        this.code = code;
        this.name = name;
        this.expression = expression;
        this.domainId = domainId;
        this.returnType = "number";
        this.status = "draft";
    }
}
